# Assignment 1 of 4

____________________________________________
**RUBRICS**
- Correct value.
- Number of statistics computed.
- Any added comment on the computed value.

## Exercise 1a: Download the dataset 
https://github.com/johnhabere/datasets/blob/main/exchange-rates_uga.csv

From the dataset accomplish the various statistics of the exchange rates using the rates (referred to as 'values' in the dataset) in the following exercises.

#### Importing the data after downloading and doing some EDA to understand it before answering the question

In [1]:
# importing the required packages
import pandas as pd

data = pd.read_csv('exchange-rates_uga.csv')
data.columns  # confirming is the column Value does exist in the dataframe

Index(['Iso3', 'StartDate', 'EndDate', 'Area Code', 'Area Code (M49)', 'Area',
       'Element Code', 'Element', 'ISO Currency Code', 'Currency', 'Year Code',
       'Year', 'Months Code', 'Months', 'Unit', 'Value', 'Flag'],
      dtype='object')

In [54]:
data.head(20)

Iso3    StartDate     EndDate  Area Code Area Code (M49)  \
0   #country+code  #date+start   #date+end        NaN             NaN   
1             UGA   1987-01-01  1987-12-31      226.0            '800   
2             UGA   1987-01-01  1987-01-31      226.0            '800   
3             UGA   1987-02-01  1987-02-28      226.0            '800   
4             UGA   1987-03-01  1987-03-31      226.0            '800   
5             UGA   1987-04-01  1987-04-30      226.0            '800   
6             UGA   1987-05-01  1987-05-31      226.0            '800   
7             UGA   1987-06-01  1987-06-30      226.0            '800   
8             UGA   1987-07-01  1987-07-31      226.0            '800   
9             UGA   1987-08-01  1987-08-31      226.0            '800   
10            UGA   1987-09-01  1987-09-30      226.0            '800   
11            UGA   1987-10-01  1987-10-31      226.0            '800   
12            UGA   1987-11-01  1987-11-30      226.0            '800   
13            UGA   1987-12-01  1987-12-31      226.0            '800   
14            UGA   1988-01-01  1988-12-31      226.0            '800   
15            UGA   1988-01-01  1988-01-31      226.0            '800   
16            UGA   1988-02-01  1988-02-29      226.0            '800   
17            UGA   1988-03-01  1988-03-31      226.0            '800   
18            UGA   1988-04-01  1988-04-30      226.0            '800   
19            UGA   1988-05-01  1988-05-31      226.0            '800   

             Area Element Code                       Element  \
0   #country+name          NaN                           NaN   
1          Uganda          LCU  Local currency units per USD   
2          Uganda          LCU  Local currency units per USD   
3          Uganda          LCU  Local currency units per USD   
4          Uganda          LCU  Local currency units per USD   
5          Uganda          LCU  Local currency units per USD   
6          Uganda          LCU  Local currency units per USD   
7          Uganda          LCU  Local currency units per USD   
8          Uganda          LCU  Local currency units per USD   
9          Uganda          LCU  Local currency units per USD   
10         Uganda          LCU  Local currency units per USD   
11         Uganda          LCU  Local currency units per USD   
12         Uganda          LCU  Local currency units per USD   
13         Uganda          LCU  Local currency units per USD   
14         Uganda          LCU  Local currency units per USD   
15         Uganda          LCU  Local currency units per USD   
16         Uganda          LCU  Local currency units per USD   
17         Uganda          LCU  Local currency units per USD   
18         Uganda          LCU  Local currency units per USD   
19         Uganda          LCU  Local currency units per USD   

   ISO Currency Code         Currency  Year Code        Year  Months Code  \
0                NaN              NaN        NaN  #date+year          NaN   
1                UGX  Uganda Shilling     1987.0        1987       7021.0   
2                UGX  Uganda Shilling     1987.0        1987       7001.0   
3                UGX  Uganda Shilling     1987.0        1987       7002.0   
4                UGX  Uganda Shilling     1987.0        1987       7003.0   
5                UGX  Uganda Shilling     1987.0        1987       7004.0   
6                UGX  Uganda Shilling     1987.0        1987       7005.0   
7                UGX  Uganda Shilling     1987.0        1987       7006.0   
8                UGX  Uganda Shilling     1987.0        1987       7007.0   
9                UGX  Uganda Shilling     1987.0        1987       7008.0   
10               UGX  Uganda Shilling     1987.0        1987       7009.0   
11               UGX  Uganda Shilling     1987.0        1987       7010.0   
12               UGX  Uganda Shilling     1987.0        1987       7011.0   
13               UGX  Uganda Shilling     1987.0        

## Exercise 1b: Create a list rates from the dataset

In [3]:
# Excluding the first row since it is a header
data_new = data.iloc[1:]

# Now creating a list rate from the dataset.
rates = data_new['Value'].tolist()

# Displaying the output
print(f'A list rates: \n {rates} ')

A list rates: 
 ['42.841266666700001', '42.841266666700001', '14.000000000000000', '14.000000000000000', '14.000000000000000', '38.095199999999998', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '106.135833333300000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '60.000000000000000', '150.000000000000000', '150.000000000000000', '150.000000000000000', '150.000000000000000', '150.000000000000000', '163.630000000000000', '223.091606308100010', '164.989275697100000', '165.000000000000000', '191.660000000000000', '200.000000000000000', '200.000000000000000', '200.000000000000000', '200.000000000000000', '200.000000000000000', '200.000000000000000', '240.000000000000000', '345.449999999999990', '370.000000000000000', '428.854666666699980', '374.286000000000000', '378.199999999999990', '379.000000000000000', '379.00000000

## Calculating statistics of the exchange rates

In [4]:
# Since statistics are applied to numeric values, I will convert 'rates' to numeric values
rates_numeric = pd.to_numeric(rates, errors='coerce')

## Exercise 2: Measures of Central Tendency

### Mean

In [7]:
# Function to calculate Mean
def calculate_mean(x):
    if len(x) == 0:
        raise ValueError("List is empty, cannot calculate mean!")
    total = sum(x)
    mean = total / len(x)
    return mean

In [8]:
calculate_mean(rates_numeric)

1383.3367407705691

##### Most of the rates are grouped together around 1383.34.

### Median

In [9]:
# Function to calculate Median
def calculate_median(x):
    if len(x) == 0:
        raise ValueError("List is empty, cannot calculate median!")
    sorted_x = sorted(x)
    n = len(sorted_x)

    if n % 2 == 0:
        # For an even-sized list, average the middle two elements
        middle_left = sorted_x[n // 2 - 1]
        middle_right = sorted_x[n // 2]
        median = (middle_left + middle_right) / 2
    else:
        # For an odd-sized list, return the middle element
        median = sorted_x[n // 2]
    return median


In [10]:
calculate_median(rates_numeric)

1196.34

##### With a median value of 1196.34, half of the data points are below and half are above this figure. This acts as a central reference point, displaying the values' center position and illuminating the central tendency of the distribution.

### Mode

In [51]:
from collections import Counter


# Function to calculate Mode
def calculate_mode(x):
    if len(x) == 0:
        raise ValueError("List is empty, cannot calculate mode!")
    # Use Counter to count occurrences of each element
    counts = Counter(x)

    # Find the maximum count
    max_count = max(counts.values())

    # Get all elements with maximum count (multiple modes are possible)
    modes = [key for key, value in counts.items() if value == max_count]

    if len(modes) == 1:
        # If there's only one mode, return it
        return modes[0]
    else:
        # If there are multiple modes, return a list of modes
        return modes

In [14]:
calculate_mode(rates_numeric)

[60.0,
 7.14300001,
 7.14299959,
 7.143,
 7.02145,
 7.13595,
 8.73533012,
 9.75160896,
 16.6133783,
 21.3914565,
 33.57103329,
 59.47877857,
 93.83923713,
 118.3045879,
 165.14839406,
 266.72679284,
 672.02,
 1400.0]

In [16]:

import statistics

# Use statistics.mode to find the mode
mode_result = statistics.mode(rates_numeric)
mode_result

60.0

## Exercise 2: Measures of Dispersion or Variability

Remember to use Bessel's correction.

### Range

In [17]:
# Function to calculate Range
def calculate_range(x):
    if len(x) == 0:
        raise ValueError("List is empty, cannot calculate range!")
    min_value = min(x)
    max_value = max(x)
    data_range = max_value - min_value
    return data_range

In [18]:
calculate_range(rates_numeric)

3840.40653550001

##### The range of extremes within the rates is demonstrated by the least value of 0.0702 and the largest value of 3840.41, highlighting the diversity of the observed rates. Therefore, the range rate of 3840.41 highlights the wide range of the dataset even more.

### Variance

In [20]:
def calculate_variance(i):
    if len(i) < 2:
        raise ValueError("Variance requires at least two data points.")
    mean = sum(i) / len(i)
    squared_diff = [(x - mean) ** 2 for x in i]
    variance = sum(squared_diff) / (
            len(data) - 1)  # Use (n-1) for sample variance, use n for population variance
    return variance

In [21]:
calculate_variance(rates_numeric)

1564355.1767827822

##### With a variance of 1564355.18, the results are significantly spread out  from the mean. This metric illustrates how far each value deviates from the central tendency. It is the average squared deviation of each data point from the mean. Practically speaking, a lower variance denotes a distribution that is more grouped around the mean, whereas a higher variance points to more diversity or variability within the dataset.

### Standard Deviation

In [55]:
# Function to calculate standard deviation
def calculate_standard_deviation(y):
    if len(y) == 0:
        raise ValueError("List is empty, cannot calculate standard deviation!")
    mean_y = calculate_mean(y)
    squared_diff_y = [(i - mean_y) ** 2 for i in y]
    variance = sum(squared_diff_y) / len(y)
    std_dev_y = variance ** 0.5
    return std_dev_y

In [56]:
calculate_standard_deviation(rates_numeric)

1250.7418505762018

##### the substantial variability and dispersion of 1250.74 indicates a broad range of values deviating from the mean.

### Interquartile Range (IQR)

In [57]:
def calculate_iqr(i):
    if len(i) < 4:
        raise ValueError("IQR requires at least four data points.")
    sorted_data = sorted(i)

    # Calculate Q1 and Q3
    q1_index = (len(sorted_data) - 1) // 4
    q3_index = 3 * (len(sorted_data) - 1) // 4

    q1 = (sorted_data[q1_index] + sorted_data[q1_index + 1]) / 2
    q3 = (sorted_data[q3_index] + sorted_data[q3_index + 1]) / 2

    # Calculate IQR
    iqr = q3 - q1
    return iqr

In [58]:
calculate_iqr(rates_numeric)

2004.7270968620803

##### The Interquartile Range (IQR), with a value of 2004.7, indicates that the middle 50% of the dataset has a spread or range of roughly 2004.73 units. This suggests that the middle 50% of the data encompass a wide range of values, with the central section of the data being moderately spread.


## Exercise 3: Calculating more statistics

- Coefficient of Variation
- quartile coefficent of dispersion
- covariance
- Pearson correlation coefficient ($\rho$)
- Percentiles (10th, 25th, 50th, 75th etc)
- Skewness
- **ADD ANY OTHER STATISTICS (THE MORE COMPUTATIONS, THE MORE POINTS)**

#### Coefficient of Variation:

In [59]:
def calculate_coefficient_of_variation(i):
    if len(i) == 0:
        raise ValueError("List is empty, cannot calculate CV!")
    mean_value = sum(i) / len(i)
    std_dev = (sum((x - mean_value) ** 2 for x in i) / len(i)) ** 0.5
    coefficient_of_variation = (std_dev / mean_value) * 100
    return coefficient_of_variation

In [60]:
calculate_coefficient_of_variation(rates_numeric)

90.41485082507769

##### A significant degree of variability in relation to the mean is indicated by a coefficient of variation value of 90.41

#### Quartile Coefficient of Dispersion

In [61]:
def calculate_quartile_coefficient_of_dispersion(i):
    if len(i) < 4:
        raise ValueError("QCD requires at least four data points.")
    sorted_data = sorted(i)
    q1_index = (len(sorted_data) - 1) // 4
    q3_index = 3 * (len(sorted_data) - 1) // 4
    q1 = (sorted_data[q1_index] + sorted_data[q1_index + 1]) / 2
    q3 = (sorted_data[q3_index] + sorted_data[q3_index + 1]) / 2
    qcd = (q3 - q1) / (q3 + q1)
    return qcd

In [62]:
calculate_quartile_coefficient_of_dispersion(rates_numeric)

0.9143960968116198

##### The gap between the first and third quartiles is less than the entire range of the data (IQR < Range), as indicated by the QCD value of 0.914

#### Covariance:

#### This should be between 2 vars thus will extract year codes to perform this statistic: i.e., fiinding a convariance between rates and area code

In [76]:
dates_coded = data_new['StartDate']

df = pd.DataFrame(dates_coded)

# Convert 'date' to categorical variable
df['date_categorical'], _ = pd.factorize(df['StartDate'])

# Now creating a list rate from the dataset.
coded_dates = df['date_categorical'].tolist()

# Displaying the output
print(f'A list dates: \n {coded_dates} ')

A list dates: 
 [0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203

In [77]:
# Since statistics are applied to numeric values, I will convert 'rates' to numeric values
coded_dates_numeric = pd.to_numeric(coded_dates, errors='coerce')

In [30]:
def calculate_covariance(k, l):
    if len(k) != len(l) or len(k) < 2:
        raise ValueError(
            "Input data should have the same length and at least two data points.")
    mean_k = sum(k) / len(k)
    mean_l = sum(l) / len(l)
    covariance = sum((k[i] - mean_k) * (l[i] - mean_l) for i in
                     range(len(k))) / (len(k) - 1)
    return covariance

In [78]:
calculate_covariance(rates_numeric, coded_dates_numeric)

-29452.270262036764

##### With a covariance value of about -29452.27, the two variables are negatively correlated.

#### Pearson Correlation Coefficient:

In [52]:
def calculate_pearson_correlation(k, l):
    if len(k) != len(l) or len(k) < 2:
        raise ValueError(
            "Input data should have the same length and at least two data points.")

    covariance = calculate_covariance(k, l)
    std_dev_k = (sum((x - sum(k) / len(k)) ** 2 for x in k) / len
    (k)) ** 0.5
    std_dev_l = (sum((x - sum(l) / len(l)) ** 2 for x in l) / len
    (l)) ** 0.5
    pearson_correlation = covariance / (std_dev_k * std_dev_l)
    return pearson_correlation

In [79]:
calculate_pearson_correlation(rates_numeric, coded_dates_numeric)

-0.12720783272405672

##### There appears to be a slight negative linear relationship between dates  and rates, as indicated by the Pearson correlation coefficient of roughly -0.13.

#### Percentiles:

In [80]:
def calculate_percentile(i, percentile):
    if len(i) < 1 or percentile < 0 or percentile > 100:
        raise ValueError("Invalid input data or percentile value.")

    sorted_i = sorted(i)
    index = (percentile / 100) * (len(sorted_i) - 1)

    if index.is_integer():
        return sorted_i[int(index)]
    else:
        lower_index = int(index // 1)
        upper_index = lower_index + 1
        lower_value = sorted_i[lower_index]
        upper_value = sorted_i[upper_index]
        return lower_value + (index - lower_index) * (upper_value - lower_value)

In [81]:
tenth_percentile = calculate_percentile(rates_numeric, 10)
second_percentile = calculate_percentile(rates_numeric, 25)
third_percentile = calculate_percentile(rates_numeric, 50)
fourth_percentile = calculate_percentile(rates_numeric, 75)

print(f' 10th Percentile: {tenth_percentile}\n 25th Percentile: '
      f'{second_percentile}\n 50th Percentile: '
      f'{third_percentile}\n 75th Percentile: {fourth_percentile}\n')

 10th Percentile: 7.14300001
 25th Percentile: 93.83923713
 50th Percentile: 1196.34
 75th Percentile: 2098.56633399208



#### I have observed that the 50th Percentile is equivalent to the median

#### Skewness

In [49]:
def calculate_skewness(i):
    if len(i) < 3:
        raise ValueError("Skewness requires at least three data points.")

    mean_value = sum(i) / len(i)
    std_dev = (sum((x - mean_value) ** 2 for x in i) / len(i)) ** 0.5
    skewness = sum(((x - mean_value) / std_dev) ** 3 for x in i) * \
               (len(i) / ((len(i) - 1) * (len(i) - 2)))
    return skewness

In [50]:
calculate_skewness(rates_numeric)

0.541059764569046

##### The data distribution has a somewhat positive skewness, as indicated by the skewness value of 0.541